pkg/aspect/types.go

```diff
@@ -0,0 +1,107 @@
+type (
+
+	// Aspect -- User visible cross cutting concern
+	// istio/metrics, istio/accessControl, istio/accessLog are all aspects of a deployment
+	// This interface is extended by aspects;
+	// ex: listChecker.Aspect adds --> CheckList(Symbol string) (bool, error)
+	// The extended interface is implemented by adapter impls.
+	// ex: //adapter/ipListChecker aspectState implements the listChecker.Aspect.
+	Aspect interface {
+		io.Closer
+		// ImplName returns the official name of adapter implementation ex. "istio/statsd"
+		ImplName() string
+	}
+
+	// Config (aspect.Config) is the internal struct used for the Aspect proto
+	// at present it only has an additional field that stores the converted
+	// and typed proto
+	Config struct {
+		istioconfig.Aspect
+		// TypedParams points to the proto after google_protobuf.Struct is converted
+		TypedParams proto.Message
+	}
+
+
+	// Adapter represents an Aspect Builder that constructs instances a specific aspect.
+	// This interface is extended by  aspects; (metrics, listChecker, ...)
+	// The extended interface is implemented by adapter implementations
+	Adapter interface {
+		io.Closer
+		// Name returns the official name of this adapter. ex. "istio/statsd".
+		Name() string
+		// Description returns a user-friendly description of this adapter.
+		Description() string
+		// DefaultConfig returns a default configuration struct for this
+		// adapter Implmentation. This will be used by the configuration system to establish
+		// the shape of the block of configuration state passed to the NewAspect method.
+		DefaultConfig() (implConfig proto.Message)
+		// ValidateConfig determines whether the given configuration meets all correctness requirements.
+		ValidateConfig(implConfig proto.Message) error
+	}
+
+	// AdapterConfig (aspect.AdapterConfig) is the internal struct used for the Adapter proto
+	// at present it only has an additional field that stores the converted
+	// and typed Args
+	AdapterConfig struct {
+		istioconfig.Adapter
+		// TypedArgs points to the proto after google_protobuf.Struct is converted
+		TypedArgs proto.Message
+	}
+
+	// CombinedConfig combines all configuration related to an aspect
+	CombinedConfig struct {
+		Aspect  *Config
+		Adapter *AdapterConfig
+	}
+
+	// Output from the Aspect Manager
+	Output struct {
+		// status code
+		Code code.Code
+		//TODO attribute mutator
+		//If any attributes should change in the context for the next call
+		//context remains immutable during the call
+	}
+	// Manager manages a specific aspect and presets a uniform interface
+	// to the rest of system
+	Manager interface {
+		// Execute dispatches to the given aspect using  config
+		// A cached instance of Aspect is provided that was previously obtained by
+		// calling NewAspect
+		// The evaluation is done under the context of an attribute bag and using
+		// an expression evaluator
+		Execute(cfg *CombinedConfig, asp Aspect, attrs attribute.Bag, mapper expr.Evaluator) (*Output, error)
```

Comment:
look at listChecker.manager's Execute and NewAspect methods.

Where to use the configuration is a choice for the aspect manager. 

The aspect manager can act on as much aspect specific config it wants to and then only use the subset of the configuration to call NewAspect.

In the above example, Both "checkAttribute" and "blackList" from 
`listcheckerpb "istio.io/api/istio/config/v1/aspect/listChecker"` are fully processed by aspect manager. So you just don't want to send down the protobuf config in its entirety. 

Think about what happens when both aspect manager *and* adapterImpl process the `blacklist` flag. They will end up negating outputs 2 times.